### ISSUE TITLE
Counterfactual search using gradient-free optimization

### ISSUE THREAD

**Issue by jklaise** [SeldonIO/alibi#176]
Currently the optimization for counterfactual search uses gradient descent both in the white-box (for TF and keras models) and black-box (using numerical gradients) case. However, this is far from ideal for models/prediction functions which are not inherently differentiable, e.g. tree ensembles. In this case one could use more appropriate gradient-free methods (e.g. CMA-ES).

This is a long term issue as it is not clear how the code should be structured to easily factor out different optimizers.

**Comment by jklaise** [comment]
This paper uses a genetic algorithm: https://arxiv.org/abs/2004.11165

**Comment by jimbudarz** [comment]
It seems that this is an issue for CEMs as well, as the numerical gradients for both often fail to find a suitable Pertinent Negatives in particular.

**Comment by jklaise** [comment]
@jimbudarz that's correct. We're currently doing some research on gradient-free methods for this. This is very much research in progress so we can't promise an integration in the near future, but we do have @RobertSamoilescu working on this as an internship project, so we hope to make some progress!

**Comment by jklaise** [comment]
Closing as this is solved by #457.
